client/state/data-layer/README.md

```diff
@@ -0,0 +1,109 @@
+Unfortunately Redux leaves a gap in terms of asynchronous data updates.
+`redux-thunk` brought us a way to separate those asynchronous pieces from our components and live within the Redux ecosystem.
+
+On the bright side, `redux-thunk` solves our needs as they relate to components.
+Individual components need only call `this.props.fetchSplines()` if they need a list of splines.
+That action creator can live in a more centralized location and is easier to reuse across the application.
+
+Unfortunately this doesn't solve the Redux side of our problems and doesn't give us what we need in terms of flexibility in the data layer to handle other goals such as offline abilities.
+It raises another ideological problem: it replaces serializable data structures describing intended changes with non-inspectable function objects.
+There is no way of introspecting those thunks as they pass through the Redux middleware and hit the reducers.
+Their intentions are opaque.
+
+Further, we are still left in a situation where our data synchronization is decentralized and redundant.
+Each thunk action creator must dispatch the API requests and handle the possible network errors, which inevitably leads to the same kind of duplication on approximately-equal-while-implemented-slightly-differently functions.
+
+## Goals and characteristics
+
+This data layer opens up our ability to make Calypso perform well offline, lessens the burden of keeping related data in sync, and provides mechanisms to reduce mobile battery consumption and minimize network bandwidth.
+
+Components should be thinking of _what_ their data needs are more than they should be thinking about _how_ to meet them.
+They should _trust_ the framework to make sure those needs are met if indeed it is possible to meet them.
+
+In contrast to the prior solutions the Redux actions which components will end up dispatching do not directly invoke the associated API calls and asynchronous fetches.
+Middleware at this layer will trap those actions and interpret them however is most fitting.
+For example, it may determine that an update to one type of data justifies also updating another related type of data at the same time.
+It may determine that some actions, such as submitting a post or comment, justify waiting a short period of time before actually requesting an update on the server to allow for cancelling accidental submissions.
+
+This middleware should break the tight coupling between Calypso's internal data structure and that presented by the APIs or backends which supply it.
+Components should ideally not need to know that some parts of related data (such as a site's settings or plan subscriptions) comes from distinct API endpoints.
+Instead the components should be able to request that they require such data the the middleware will intercept the request to fulfill it accordingly.
+
+## Implementation
+
+The data layer _intercepts_ Redux actions.
+Once it does this the action will be dropped entirely.
```

Comment:
Reading this, I was worried that two API middlewares couldn't intercept the same action, though looking at the implementation it seems this should work fine. Maybe clarify wording?